rewrite-java/src/main/java/org/openrewrite/java/cleanup/XmlParserXXEVulnerabilityFix.java

```diff
@@ -0,0 +1,153 @@
+    private static final String XML_PARSER_INITIALIZATION_METHOD = "xml-parser-initialization-method";
+    private static final String XML_FACTORY_VARIABLE_NAME = "xml-factory-variable-name";
+
+    private static final ThreadLocal<JavaParser> JAVA_PARSER_THREAD_LOCAL = ThreadLocal.withInitial(() -> JavaParser.fromJavaVersion().build());
+
+    @Override
+    public String getDisplayName() {
+        return "XML Parser XXE Vulnerability Fix.";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Avoid exposing dangerous features of the XML parser by setting XMLInputFactory IS_SUPPORTING_EXTERNAL_ENTITIES and SUPPORT_DTD properties to false.";
+    }
+
+    @Override
+    protected @Nullable TreeVisitor<?, ExecutionContext> getApplicableTest() {
+        return new HasTypes(Collections.singletonList(XML_FACTORY_FQN)).getVisitor();
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, ExecutionContext executionContext) {
+                J.ClassDeclaration cd = super.visitClassDeclaration(classDecl, executionContext);
+                Cursor supportsExternalCursor = getCursor().getMessage(SUPPORTING_EXTERNAL_ENTITIES_PROPERTY_NAME);
+                Cursor supportsDTDCursor = getCursor().getMessage(SUPPORT_DTD_PROPERTY_NAME);
+                Cursor initializationCursor = getCursor().getMessage(XML_PARSER_INITIALIZATION_METHOD);
+                String xmlFactoryVariableName = getCursor().getMessage(XML_FACTORY_VARIABLE_NAME);
+
+                Cursor setPropertyBlockCursor = null;
+                if (supportsExternalCursor == null && supportsDTDCursor == null) {
+                    setPropertyBlockCursor = initializationCursor;
+                } else if (supportsExternalCursor == null ^ supportsDTDCursor == null) {
+                    setPropertyBlockCursor = supportsExternalCursor == null ? supportsDTDCursor : supportsExternalCursor;
+                }
+                if (setPropertyBlockCursor != null && xmlFactoryVariableName != null) {
+                    doAfterVisit(new XmlFactoryInsertPropertyStatementVisitor(setPropertyBlockCursor.getValue(), xmlFactoryVariableName,supportsExternalCursor == null, supportsDTDCursor == null));
+                }
+                return cd;
+            }
+
+            @Override
+            public J.VariableDeclarations.NamedVariable visitVariable(J.VariableDeclarations.NamedVariable variable, ExecutionContext executionContext) {
+                J.VariableDeclarations.NamedVariable v = super.visitVariable(variable, executionContext);
+                if (v.getType() != null && v.getType() instanceof JavaType.Class
+                        && XML_FACTORY_FQN.equals(((JavaType.Class) v.getType()).getFullyQualifiedName())) {
```

Comment:
Instead of checking for `JavaType.Class` check for `JavaType.FullyQualified`. It's always possible as an implementation detail you might be looking at an AST that has a `JavaType.ShallowClass` instead of a `JavaType.Class`